### ISSUE TITLE
Passing `c_args`  to a subproject does not work

### ISSUE THREAD

**Issue by astro-stan** [mesonbuild/meson#13260]
**Describe the bug**
I don't see anything in the docs to suggest this is not possible, so I am assuming this is a bug. Let me know if I misunderstood something.

I am trying to pass a string with build args to my subproject. But when I `get_option('c_args')` on the subproject's side I get an empty list.

**To Reproduce**
Here is a minimal example to reproduce:

* Root file:
```meson
# ./meson.build

project('proj', 'c',
  version : '0.1',
  subproject_dir: 'subprojects',
)

subproject('sub1', default_options:{
  'c_args': '-Dtest'
})
```

* Subproject:

```meson
# ./subprojects/sub1/meson.build

project('sub1', 'c',
  version : '0.1')

message(get_option('c_args'))
```

**Expected behavior**
I would expect to see the options propagate to the subproject and be applied to all build targets.

This is the case when I pass them directly to the subproject like so:

```sh
./subprojects/sub1$ meson build -Dc_args="-Dtest"
```

But not when passed via the main project.

**system parameters**
* Is this a [cross build](https://mesonbuild.com/Cross-compilation.html) or just a plain native build (for the same computer)? Native
* what operating system (e.g. MacOS Catalina, Windows 10, CentOS 8.0, Ubuntu 18.04, etc.): Linux
* what Python version are you using e.g. 3.8.0: 3.11
* what `meson --version`: 1.4.0
* what `ninja --version` if it's a Ninja build: 1.10.1

**Comment by eyalitki** [comment]
TL;DR
-
Loading is done in stages, yet caching causes the language args to be **skipped** for subprojects, hence ```c_args``` isn't covered. Will now start working on a patch.

Full Triage:
-
In short, the handling of default options is done in stages, and the handling of "compiler" options is deferred to the later load of the compiler - [source](https://github.com/mesonbuild/meson/blob/master/mesonbuild/coredata.py#L704). Despite the code check itself, this isn't fully accurate, as the comment is more accurate:
```
# Skip base, compiler, and backend options, they are handled when
# adding languages and setting backend.
if self.optstore.is_compiler_option(k) or self.optstore.is_backend_option(k):
```

Thing is, language args (of the form ```<language>_<option>```) are only added when the language is detected and the compiler is detected for it. That said, there is a caching of compiler for language, hence the subproject won't go through this discovery process. This causes language args to only load **once** for the main project, while the subproject will only load the "real" compiler options instead of also the language options.

An example trace with added debug prints (trace taken from reproduce with add "-Dtest2" as default options for the main project):
```
The Meson build system
Version: 1.6.99
Source dir: /home/eyalitki/tests/subproject_c_flags
Build dir: /tmp/build
Build type: native build
EI: set_default_options() - c_args = -Dtest2
EI: set_default_options() - skipping compiler/backend options
Project name: proj
Project version: 0.1
EI: add_languages_for()
EI: add_languages_for() - No compiler yet, invoke detect_compiler_for()
EI: detect_compiler_for() - c,
EI: detect_c_compiler()
EI: _detect_c_or_cpp_compiler()
EI: add_lang_args() - c_args = ['-Dtest2']
EI: add_lang_args() - c_link_args = []
EI: add_compiler_options() - c_std
C compiler for the host machine: cc (gcc 13.2.0 "cc (Ubuntu 13.2.0-23ubuntu4) 13.2.0")
C linker for the host machine: cc ld.bfd 2.42
EI: add_languages_for()
EI: add_languages_for() - No compiler yet, invoke detect_compiler_for()
EI: detect_compiler_for() - c,
EI: detect_c_compiler()
EI: _detect_c_or_cpp_compiler()
EI: add_lang_args() - build.c_args = []
EI: add_lang_args() - build.c_link_args = []
EI: add_compiler_options() - build.c_std
Host machine cpu family: x86_64
Host machine cpu: x86_64

Executing subproject sub1

EI: set_default_options() - sub1:c_args = -Dtest
EI: set_default_options() - skipping compiler/backend options
EI: set_default_options() - c_args = -Dtest2
EI: set_default_options() - skipping compiler/backend options
sub1| Project name: sub1
sub1| Project version: 0.1
EI: add_languages_for()
EI: add_languages_for() - Compiler is found, invoke process_compiler_options()
EI: add_compiler_options() - c_std
EI: add_compiler_options() - c_std
EI: add_languages_for() - Subproject - add_compiler_options and that's it
EI: add_compiler_options() - sub1:c_std
EI: add_compiler_options() - sub1:c_std
sub1| C compiler for the host machine: cc (gcc 13.2.0 "cc (Ubuntu 13.2.0-23ubuntu4) 13.2.0")
sub1| C linker for the host machine: cc ld.bfd 2.42
EI: add_languages_for()
EI: add_languages_for() - Compiler is found, invoke process_compiler_options()
EI: add_compiler_options() - build.c_std
EI: add_compiler_options() - build.c_std
EI: add_languages_for() - Subproject - add_compiler_options and that's it
EI: add_compiler_options() - sub1:build.c_std
EI: add_compiler_options() - sub1:build.c_std
sub1| Message: ['-Dtest2']
sub1| Build targets in project: 0
sub1| Subproject sub1 finished.

Build targets in project: 0

proj 0.1

  Subprojects
    sub1: YES

Found ninja-1.11.1 at /usr/bin/ninja
```

**Comment by bonzini** [comment]
@jpakkane is this related to the option refactoring?

**Comment by jpakkane** [comment]
Not really, but this works now. I tried the steps above and the output is the following:

```
...
sub1| Message: ['-Dtest']
...
```

And `meson configure` says:

```
...
Currently set option augments:
sub1:c_args          -Dtest    
'''
```

**Comment by Woynert** [comment]
Hi, I tested this on meson 1.10.1 and it seems c_args gets correctly set on the subproject like @jpakkane   shows, however it still isn't applied to the build target:

```
project('my_lib', 'c')

message(get_option('c_args')) # prints ['-Dtest'] OK.

my_lib = library('mylib', './mylib.c', include_directories: ...) # c_args ignored
my_dep = declare_dependency ...
EOF
```

I had to manually set it like this for it to work: 

```
my_lib = library('mylib',
  './mylib.c',
  c_args: get_option('c_args'), # <--
  include_directories: ...
)
```

Worth reopening?

EDIT: [minimal repro project](https://github.com/Woynert/meson-minimal-repro1).
